src/parser.rs

```diff
@@ -0,0 +1,435 @@
+        let identifier = match self.pop()? {
+            Token::AlphaNumeric(identifier) => {
+                // TODO: Borrow?
+                Identifier::AlphaNumeric(identifier.to_string())
+            }
+            Token::Numeric(n) => Identifier::Numeric(n),
+            tok => return Err(UnexpectedToken(tok)),
+        };
+
+        Ok(identifier)
+    }
+
+    /// Parse all pre-release identifiers, separated by dots.
+    ///
+    /// Like, `abcdef.1234`.
+    fn pre(&mut self) -> Result<Vec<Identifier>, Error<'input>> {
+        match self.peek() {
+            Some(&Token::Hyphen) => {}
+            _ => return Ok(vec![]),
+        }
+
+        // pop the peeked hyphen.
+        self.pop()?;
+        self.parts()
+    }
+
+    /// Parse a dot-separated set of identifiers.
+    fn parts(&mut self) -> Result<Vec<Identifier>, Error<'input>> {
+        let mut parts = Vec::new();
+
+        parts.push(self.identifier()?);
+
+        loop {
+            match self.peek() {
+                Some(&Token::Dot) => {}
+                _ => break,
+            }
+
+            // pop the peeked hyphen.
+            self.pop()?;
+
+            parts.push(self.identifier()?);
+        }
+
+        Ok(parts)
+    }
+
+    /// Parse optional build metadata.
+    ///
+    /// Like, `` (empty), or `+abcdef`.
+    fn plus_build_metadata(&mut self) -> Result<Vec<Identifier>, Error<'input>> {
+        match self.peek() {
+            Some(&Token::Plus) => {}
+            _ => return Ok(vec![]),
+        }
+
+        // pop the plus.
+        self.pop()?;
+        self.parts()
+    }
+
+    /// Optionally parse a single operator.
+    ///
+    /// Like, `~`, or `^`.
+    pub fn op(&mut self) -> Result<Op, Error<'input>> {
+        use self::Token::*;
+
+        let op = match self.peek() {
+            Some(&Eq) => Op::Ex,
+            Some(&Gt) => Op::Gt,
+            Some(&GtEq) => Op::GtEq,
+            Some(&Lt) => Op::Lt,
+            Some(&LtEq) => Op::LtEq,
+            Some(&Tilde) => Op::Tilde,
+            Some(&Caret) => Op::Compatible,
+            // default op
+            _ => return Ok(Op::Compatible),
+        };
+
+        // remove the matched token.
+        self.pop()?;
+        self.skip_whitespace()?;
+        Ok(op)
+    }
+
+    /// Parse a single predicate.
+    ///
+    /// Like, `^1`, or `>=2.0.0`.
+    pub fn predicate(&mut self) -> Result<Option<Predicate>, Error<'input>> {
+        // empty predicate, treated the same as wildcard.
+        if self.peek().is_none() {
+            return Ok(None);
+        }
+
+        let mut op = self.op()?;
+
+        let major = match self.component()? {
+            Some(major) => major,
+            None => return Ok(None),
+        };
+
+        let (minor, minor_wildcard) = self.dot_component()?;
+        let (patch, patch_wildcard) = self.dot_component()?;
+        let pre = self.pre()?;
+
+        // TODO: avoid illegal combinations, like `1.*.0`.
```

Comment:
when we merge this, we should make issues for each of these TODOs

Reply:
This is something we currently accept, and I added a test for it in #35 to make sure we stay consistent. Not sure if we should break it or not.

Reply:
i think https://github.com/steveklabnik/semver-parser/issues/36 should be the answer to this question, basically

Reply:
so, using https://semver.npmjs.com/, it seems that this is accepted. guess we should keep it

Reply:
Yeah, it's just a bit unsound. The following requirements are currently equivalent as per our and their implementation:
* `1.*`
* `1.*.*`
* `1.*.0`, `1.*.1`, ... any number. Note how they are ignored.

Maybe this should generate a warning? A user might expect that `1.*.10` matches any 
patch-release `10` across different minor releases.